Add documentation about new ScyllaCluster's exposing options (#1432)
scylladb/scylla-operator: docs/source/exposing.md

# Exposing ScyllaCluster

This document explains how ScyllaDB Operator exposes ScyllaClusters in different network setups.
A ScyllaCluster can be exposed in various network configurations, independently to clients and nodes.

``` note::
ScyllaClusters can be only exposed when the ScyllaDB version used version is `>=2023.1` ScyllaDB Enterprise or `>=5.2` ScyllaDB Open Source.
```

## Expose Options

> `exposeOptions` as name suggests consists of configuration options for exposing ScyllaCluster's.

Reviewer: ```suggestion
`exposeOptions` specifies configuration options for exposing ScyllaClusters.
```
I'd drop the "as name suggests", comes of as a bit aggressive :stuck_out_tongue:

> A ScyllaCluster created without any `exposeOptions` is equivalent to the following:

Reviewer: ```suggestion
A ScyllaCluster created without any `exposeOptions` specified takes on the following default values:
```
nit

Author: I like submitted one better, it was also suggested to call it this way in other remark.

```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
     type: ClusterIP
    broadcastOptions:
      clients:
        type: ServiceClusterIP
      nodes:
        type: ServiceClusterIP
```

> The following sections cover what every field controls and what are the configuration options are.

Reviewer: ```suggestion
The following sections cover what every field controls and what the configuration options are.
```

Author: done

### Node Service Template

`nodeService` serves as a template for a node-dedicated Service managed by the Scylla Operator for each node within a ScyllaCluster. 
The properties of the Services depend on the selected type. 
Additionally, there's an option to define custom annotations, incorporated into each node's Service,
which might be useful for further tweaking the Service properties or related objects.

#### Headless Type

For `Headless` type, Scylla Operator creates a Headless Service with a selector pointing to the particular node in the ScyllaCluster.
Such Service doesn't provide any additional IP addresses, and the internal DNS record resolves to the PodIP of a node.

This type of Service is useful when ScyllaCluster nodes broadcast PodIPs to clients or other nodes.

Example:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
     type: Headless
```

#### ClusterIP Type

For `ClusterIP` type, Scylla Operator creates a ClusterIP Service backed by a specific node in the ScyllaCluster.

> These IP addresses are only routable within the same Kubernetes cluster, so it's a good fit, if you don't want to expose it to other networks.

Reviewer: ```suggestion
These IP addresses are only routable within the same Kubernetes cluster, so it's a good fit if you don't want to expose them to other networks.
```

Author: done

Example:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
     type: LoadBalancer
     loadBalancerClass: my-custom-load-balancer-class
```

---

### Broadcast Options

Broadcast options control what is the source of the address being broadcasted to clients and nodes.
It's configured independently for clients and nodes because you may want to expose these two types of traffic on different networks.
Using different networks can help manage costs, reliability, latency, security policies or other metrics you care about.

#### PodIP Type

Address broadcasted to clients/nodes is taken from Pod.
By default, the address is taken from Pod's `status.PodIP` field.
Because a Pod can use multiple address, you may want to provide source options by specifying `podIP.source`.

Example:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    broadcastOptions:
       clients:
         type: PodIP
         podIP:
           source: Status
```

#### ServiceClusterIP Type

Address broadcasted to clients or nodes is taken from `spec.ClusterIP` field of a node's dedicated Service.

> In order to configure it, the `nodeService` template must specify a Service having ClusterIP.

Reviewer: ```suggestion
In order to configure it, the `nodeService` template must specify a Service having a ClusterIP assigned.
```
nit

Author: done

Example:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    broadcastOptions:
       clients:
         type: ServiceClusterIP
```

#### ServiceLoadBalancerIngress Type

Address broadcasted to clients/nodes is taken from the node dedicated Service, from `status.ingress[0].ipAddress` or `status.ingress[0].hostname` field.

In order to configure it, the `nodeService` template must specify the LoadBalancer Service.

Example:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    broadcastOptions:
       clients:
         type: ServiceLoadBalancerIngress
         podIP:
           source: Status
```

## Deployment Examples

> Following section contains several specific examples of various network scenarios and explain how nodes and clients communicate with one another.

Reviewer: ```suggestion
The following section contains several specific examples of various network scenarios and explains how nodes and clients communicate with one another.
```

Author: done

### In-cluster only

ScyllaCluster definition:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
      type: ClusterIP
    broadcastOptions:
      clients:
        type: ServiceClusterIP
      nodes:
        type: ServiceClusterIP
```

Both client and nodes are deployed within the same Kubernetes cluster. 
They talk through ClusterIP addresses taken from the Service.
Because ClusterIP Services are only routable within the same Kubernetes cluster, this cluster won't be reachable from outside.

![ClusterIPs](static/exposing/clusterip.svg)

### In-cluster node-to-node, VPC clients-to-nodes

ScyllaCluster definition:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
      type: ClusterIP
    broadcastOptions:
      clients:
        type: PodIP
      nodes:
        type: ServiceClusterIP
```

In this scenario, we assume that the Pod IP subnet is routable within a VPC. 

> Clients within the VPC network can communicate directly with ScyllaCluster nodes using Pod IPs.

Reviewer: ```suggestion
Clients within the VPC network can communicate directly with ScyllaCluster nodes using PodIPs.
```
nit

Author: done

![MultiDC](static/exposing/multidc.svg)

### Internet

ScyllaCluster definition:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
      type: LoadBalancer
    broadcastOptions:
      clients:
        type: ServiceLoadBalancerIngress
      nodes:
        type: ClusterIP 
```

We assume that a Kubernetes cluster has been deployed in a cloud provider environment that supports external load balancers. 
By specifying the LoadBalancer type in the nodeService template, the Scylla Operator generates a dedicated LB Service for each node. 
The cloud provider then establishes an external load balancer with an internet-accessible address. 
ScyllaDB nodes broadcast this external address to clients, enabling drivers to connect and discover other nodes. 
Since all ScyllaDB nodes reside within the same Kubernetes cluster, there is no need to route traffic through the internet. 
Consequently, the nodes are configured to communicate via ClusterIP, which is also accessible within LoadBalancer Services.

![Internet](static/exposing/loadbalancer.svg)

-- -

Other more complex scenarios can be built upon these simple ones.

> If you have any more questions regarding it, we will welcome you on our [ScyllaDB Users Slack](http://slack.scylladb.com/), under `#scylla-operator` channel.

Reviewer: ```suggestion
Join the `#scylla-operator` Slack Channel in [ScyllaDB Users](http://slack.scylladb.com/) for comments and assistance related to exposing ScyllaClusters.
```
nit

Reviewer: On the other hand, do we really want to have that in separate docs? We have a link pointing to the users slack in the footer?

Author: Removed

# Exposing ScyllaCluster

> This documentation explains how the Scylla Operator exposes a ScyllaCluster on different network setups.

Author: done

The ScyllaCluster can be exposed in various network configurations, independently to clients and other nodes.

:::{note}

> ScyllaCluster's can be only exposed when used ScyllaDB Enterprise version is `>=2023.1` and `>=5.2` for ScyllaDB Open Source.

Author: done

:::

## Expose Options

> The ScyllaCluster created without specifying any `exposeOptions` defaults to following `exposeOptions`:

Reviewer 2: ```suggestion
The ScyllaCluster created without specifying any `exposeOptions` is equivalent to the following `exposeOptions`:
```
(to make sure it covers ScyllaClusters created before this change which don't technically default on the API level.)

> The ScyllaCluster created without specifying any `exposeOptions` defaults to following `exposeOptions`:

Author: done

> The ScyllaCluster created without specifying any `exposeOptions` defaults to following `exposeOptions`:

Author: done

```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
     type: ClusterIP
    broadcastOptions:
      clients:
        type: ServiceClusterIP
      nodes:
        type: ServiceClusterIP
```

> In following sections, we will cover what every field controls and what are the configuration options.

Author: done

### Node Service Template

> nodeService serves as a template for a node-dedicated Service deployed by the Scylla Operator for each node within a ScyllaCluster. 

Reviewer 2: ```suggestion
`nodeService` serves as a template for a node-dedicated Service managed by the Scylla Operator for each node within a ScyllaCluster. 
```

Author: done

> The properties you receive depend on the type you select. 

Author: done

> Additionally, you have the option to define custom annotations, which will be incorporated into each node Service, 

Author: done

> Additionally, you have the option to define custom annotations, which will be incorporated into each node Service, 
> these might be useful to further tweak Service resources.

Reviewer 2: ```suggestion
Additionally, you have the option to define custom annotations, which will be incorporated into each node Service.
They might be useful to further tweak Service properties or related objects.
```

> these might be useful to further tweak Service resources.

Author: done

> #### `Headless` Type

Author: done

> For `Headless` type, Scylla Operator creates a Headless Service with selector pointing to particular node in the ScyllaCluster.

Author: done

Such Service doesn't provide any additional IP address, and internal DNS records points directly to PodIP of a node.

> This type of Service is useful when ScyllaCluster nodes broadcast PodIPs to clients or other nodes.

Author: it's useful in both, what's wrong with `or` in this case?

> This type of Service is useful when ScyllaCluster nodes broadcast PodIPs to clients or other nodes.

Reviewer: if it only broadcasts PodIPs for clients or nodes exclusively, it uses some other addresses for the other - headless is not sufficient in that case? It's nitpicking anyway but I hope my point makes sense

> This type of Service is useful when ScyllaCluster nodes broadcast PodIPs to clients or other nodes.

Author: changed.

Example:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
     type: Headless
```

#### `ClusterIP` Type

> For `ClusterIP` type, Scylla Operator creates a ClusterIP Service pointing to particular node in the ScyllaCluster.

Author: done

> ClusterIP Service assigns an IP address from a pool of IP addresses that your Kubernetes cluster has reserved for that purpose.

Author: removed

These IP addresses are only routable within the same Kubernetes cluster, so it's a good fit, if you don't want to expose it to other networks.

Example:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
     type: ClusterIP
```

#### `LoadBalancer` Type

> For the 'LoadBalancer' type, the Scylla Operator generates a LoadBalancer Service that directs traffic to a specific node within the ScyllaCluster. 

Author: done

On platforms with support for external load balancers, this Service provisions one. 
The accessibility of this load balancer's address depends on the platform and any customizations made; in some cases it may be reachable from the internal network or public Internet.

> LoadBalancer Services are a superset of ClusterIP Services, implying that each LoadBalancer Service also possesses an allocated ClusterIP. 

Reviewer 2: ```suggestion
LoadBalancer Services are a superset of ClusterIP Services, implying that each LoadBalancer Service also contains an allocated ClusterIP. 
```

Author: done

> LoadBalancer Services are a superset of ClusterIP Services, implying that each LoadBalancer Service also possesses an allocated ClusterIP. 

Author: done

> These LoadBalancer Services can be configured using the following fields, and these configurations are mirrored across every node Service:

Reviewer 2: ```suggestion
They can be configured using the following fields which propagate to every node Service:
```

Author: done

* externalTrafficPolicy
* internalTrafficPolicy
* loadBalancerClass
* allocateLoadBalancerNodePorts

> Check [Kubernetes Service documentation](https://kubernetes.io/docs/concepts/services-networking/service), to know more about these options.

Author: done

Example:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
     type: LoadBalancer
     loadBalancerClass: my-custom-load-balancer-class
```

---

### Broadcast Options

> Broadcast options controls what is the source of the address being broadcasted to other clients and nodes.

Author: done

> It's configured independently for clients and nodes because you may want to expose these two types of traffic on different networks, due to costs or security concerns.

Reviewer 2: ```suggestion
It's configured independently for clients and nodes because you may want to expose these two types of traffic on different networks. Using different networks can help manage costs, reliability, latency, security policies or other metrics you care about.
```

Author: done

#### `PodIP` Type

Address broadcasted to clients/nodes is taken from Pod.

> Because Pod address may be ambiguous, you may provide what's the source by setting `podIP.source`.

Reviewer 2: ```suggestion
Because a Pod can use multiple address, you may want to provide source options by specifying `podIP.source`.
```

Author: done

> Because Pod address may be ambiguous, you may provide what's the source by setting `podIP.source`.

Author: done

> By default, address is taken from Pod `status.PodIP`.

Reviewer 2: ```suggestion
By default, the address is taken from Pod's `status.PodIP` field.
```

Author: done

> By default, address is taken from Pod `status.PodIP`.

Author: done

Example:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    broadcastOptions:
       clients:
         type: PodIP
         podIP:
           source: Status
```

#### `ServiceClusterIP` Type

> Address broadcasted to clients/nodes is taken from node dedicated Service, from `spec.ClusterIP` field.

Author: done

> In order to configure it, your `nodeService` template must deploy Service having ClusterIP.

Reviewer: ```suggestion
In order to configure it, the `nodeService` template must specify a Service of type ClusterIP.
```
or something along these lines

Author: done

Example:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    broadcastOptions:
       clients:
         type: ServiceClusterIP
```

#### `ServiceLoadBalancerIngress` Type

Address broadcasted to clients/nodes is taken from the node dedicated Service, from `status.ingress[0].ipAddress` or `status.ingress[0].hostname` field.

> In order to configure it, your `nodeService` template must deploy LoadBalancer Service.

Reviewer: I'd stick to one form and avoid addressing the users or developers directly - in other words avoid you, your, we etc. unless you want to stick with it throughout the entire document. Applies globally.

Author: done

Example:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    broadcastOptions:
       clients:
         type: ServiceLoadBalancerIngress
         podIP:
           source: Status
```

## Deployment Examples

> In this section, we will showcase specific examples of various network scenarios and explain how nodes and clients communicate with one another.

Reviewer: ditto

Author: done

> In this section, we will showcase specific examples of various network scenarios and explain how nodes and clients communicate with one another.

Reviewer 2: ```suggestion
In this section, we will showcase several specific examples of various network scenarios and explain how nodes and clients communicate with one another.
```

Author: done

### In-cluster only

ScyllaCluster definition:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
      type: ClusterIP
    broadcastOptions:
      clients:
        type: ServiceClusterIP
      nodes:
        type: ServiceClusterIP
```

Both client and nodes are deployed within the same Kubernetes cluster. 

> They talk through IP addresses from ClusterIP Service.

Author: done

> Because ClusterIP Services are only routable within the same Kubernetes cluster, connection from outside is impossible.

Reviewer 2: ```suggestion
Because ClusterIP Services are only routable within the same Kubernetes cluster, this cluster won't be reachable from outside.
```

Author: done

![ClusterIPs](static/exposing/clusterip.svg)

### In-cluster node-to-node, VPC clients-to-nodes

ScyllaCluster definition:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
      type: ClusterIP
    broadcastOptions:
      clients:
        type: PodIP
      nodes:

>         type: ServiceClusterIP

Author: Any reason why not having it? It's something user can configure, no matter what's the reason behind it. This sections showcases how traffic flows to let users better understand the outcome of every option, not to recommend how they should deploy their application.

>         type: ServiceClusterIP

Reviewer: I just think it's just somewhat confusing when compared to the next example which only uses PodIPs - having examples like these is usually meant to showcase the recommended configuration for different setups, so this sort of implies that there's something different about this scenario which calls for using ClusterIPs.

>         type: ServiceClusterIP

Author: Would you mind elaborating what is confusing about this example compared to the next one?

> having examples like these is usually meant to showcase the recommended configuration

Example is not recommendation. I haven't used "recommended" word anywhere in this document. 

>  this sort of implies that there's something different about this scenario which calls for using ClusterIPs.

isn't it a different scenario?

>         type: ServiceClusterIP

Reviewer: > Would you mind elaborating what is confusing about this example compared to the next one?

The description doesn't say anything about why you'd want to use CluserIPs for node to node communication. Which I think is confusing, especially since the first sentence in the description is "In this scenario, we assume that the Pod IP subnet is routable within a VPC.".

> Example is not recommendation. I haven't used "recommended" word anywhere in this document.

Well, given there are no other recommendations for the network setup in out docs, users are going to use them as such.

> isn't it a different scenario?

It is, but my point is there isn't anything specific about it that would explain the different setup, other than that I can *also* be set up this way.

>         type: ServiceClusterIP

Author: > The description doesn't say anything about why you'd want to use CluserIPs for node to node communication. 

Why would I describe something that I don't intent to explain? I don't need argument behind this scenario to show how traffic flows when cluster is configured this way.

> especially since the first sentence in the description is "In this scenario, we assume that the Pod IP subnet is routable within a VPC.".

Isn't it a required assumption for client-node-traffic available in this scenario?

>  my point is there isn't anything specific about it that would explain the different setup, other than that it can also be set up this way.

So you suggest to remove it? How not having anything would help understanding flows in different scenarios?

>         type: ServiceClusterIP

Reviewer: I understand your point about this only showcasing the traffic flow, but I'm pretty sure people are going to use it as a reference for how to set up their scyllaclusters in their particular network setups, that's why I think it could use some explanation. Anyway I now see that it's explained in the earlier sections, so I'm resolving this comment.

```

In this scenario, we assume that the Pod IP subnet is routable within a VPC. 
Clients within the VPC network can communicate directly with ScyllaCluster nodes using Pod IPs.
Nodes communicate with each other exclusively within the same Kubernetes cluster.

![PodIPs](static/exposing/podips.svg)

> ### Multi DC

Author: It's about showcasing how traffic flows, not how to connect two clusters. Image contains VPC's from different regions, i don't see what's wrong with it. Changed `DC` to `Region`.

> ### Multi DC

Reviewer: There's nothing wrong with it per se, but it does add extra complexity, and I'm not sure whether it introduces any additional value. All of a sudden there're now multiple clusters, which are somehow connected to each other, while the other pictures are just a single cluster + client. Maybe just putting a client in a separate, peered network would convey the same amount of information?

> ### Multi DC

Author: Again, this document isn't about connecting two clusters, it's irrelevant to how it's done. It's omitted on purpose.

> Maybe just putting a client in a separate, peered network would convey the same amount of information?

How putting client in a separate peered network would show node-to-node communication across VPCs?

--- 

Put aside comments about not bringing any value unless you have good arguments behind it.

> ScyllaCluster definition:

Reviewer 2: should we show both DC's and the externalSeeds option? or at least put a link to where multiDC is shown

Reviewer 2: maybe the whole multidc is better to be left to the port of the docs about multidc

Author: This document is about exposing options, connecting two clusters - no matter how - is not tied to it. We can add link to page describing it once we have it.

Reviewer 2: I admit I am confused by the answer.

What's the point of showing a setup of 2 ScyllaClusters in 2 Kubernetes cluster without connecting them?

this picture `![MultiDC](static/exposing/multidc.svg)` says multidc :man_shrugging:

Author: Remove "region" term in favor of multiple VPCs.

> What's the point of showing a setup of 2 ScyllaClusters in 2 Kubernetes cluster without connecting them?

I don't show how to configure clients either and it doesn't make anyone confused :man_shrugging: 

Again, describing how to connect two clusters is the not point of this document.

Reviewer 2: Is this creating 1 ScyllaCluster (in one of the Kubernetes clusters) or 2 (1 in each)? I don't see that explicitly stated, but the picture suggests 2 distinct ScyllaClusters.
Should the picture mark ScyllaCluster object boundaries?

> Nodes, deployed in these distinct VPCs,

This seems like 2 ScyllaClusters which may lead users to think that this sets up the nodes to communicate with each other, but it doesn't. Can you at least make an explicit disclamer that the nodes won't actually talk to each other without using externalSeeds when bootstraping them?

Author: > Should the picture mark ScyllaCluster object boundaries?

There's a kube cluster boundary, seems obvious these are two obvious scyllaclusters.

> This seems like 2 ScyllaClusters which may lead users to think that this sets up the nodes to communicate with each other, but it doesn't. Can you at least make an explicit disclamer that the nodes won't actually talk to each other without using externalSeeds when bootstraping them?

It's mentioned that exposeOptions only enables establishing connection. It doesn't say it establishes it. Although added explicit mention that there's a separate page about connecting two.

```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
      type: Headless
    broadcastOptions:
      clients:
        type: PodIP
      nodes:
        type: PodIP
```

In this scenario, we set up two separate Kubernetes clusters in different regions, each within its respective VPC. 
These VPCs are interconnected to facilitate inter-VPC connectivity. 
We operate on the assumption that the Pod IP subnet is routable within each VPC. 
Nodes, deployed in these distinct regions, broadcast their Pod IP addresses, enabling them to establish connections with one another. 
Clients, whether deployed within the same Kubernetes cluster or within a VPC, have the capability to reach nodes using their Pod IPs.

Since there is no requirement for any address other than the Pod IP, the 'Headless' service type is sufficient.

![MultiDC](static/exposing/multidc.svg)

### Internet

ScyllaCluster definition:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
spec:
  exposeOptions:
    nodeService:
      type: LoadBalancer
    broadcastOptions:
      clients:
        type: ServiceLoadBalancerIngress
      nodes:
        type: ClusterIP 
```

> We assume that a Kubernetes cluster has been deployed in a cloud provider environment that supports external load balancers. 

Reviewer 2: since you've already shown how to configure it generically, I wonder if these should pick a particular platform like EKS or GKE and show it really e2e, even if only for one.

Reviewer 2: although, maybe we can defer that

By specifying the LoadBalancer type in the nodeService template, the Scylla Operator generates a dedicated LB Service for each node. 
The cloud provider then establishes an external load balancer with an internet-accessible address. 
ScyllaDB nodes broadcast this external address to clients, enabling drivers to connect and discover other nodes. 
Since all ScyllaDB nodes reside within the same Kubernetes cluster, there is no need to route traffic through the internet. 
Consequently, the nodes are configured to communicate via ClusterIP, which is also accessible within LoadBalancer Services.

![Internet](static/exposing/loadbalancer.svg)

-- -

> Other more complex scenarios should be built upon these simple ones.

Reviewer 2: ```suggestion
Other more complex scenarios can be built upon these simple ones.
```

Author: done

> We hope you understand how ScyllaCluster nodes are being exposed via different `exposeOptions` configurations.

Author: done
